CONTRIBUTING.md

```diff
@@ -0,0 +1,115 @@
+
+Check which of these types are relevant and write tests for them.
+
+Most of these types will be irrelevant! Use your judgment to pick the ones that are likely to come up in a real use case.
+
+- Numbers
+  - Number type (`number`)
+  - Single literal (`1`)
+  - Union of literals (`1 | 2`)
+  - Bigint type (`bigint`)
+  - Single bigint literal (`1n`)
+  - Union of bigint literals (`1n | 2n`)
+  - Unions of numbers and bigints (`1 | 2n`)
+- Strings
+  - String type (`string`)
+  - Single literal (`"cat"`)
+  - Union of literals (`"cat" | "dog"`)
+  - Template with a type slot (`` `id_${number}` ``)
+  - Template with a literal slot (`` `id_${1 | 2}` ``)
+  - Template with multiple slots (`` `id_${1 | 2}_${3 | 4}` ``)
+- Arrays
+  - Array of a single type (`Array<number>`)
+  - Array of a union type (`Array<string | number>`)
+  - Array of literal types (`Array<"cat" | "dog">`)
+  - Union of similar arrays (`Array<1 | 2> | Array<2 | 3>`)
+  - Nested arrays (`Array<Array<number>>`)
+- Tuples
+  - Tuple of a single type (`[number, number, number]`)
+  - Tuple of different types (`[number, string, boolean]`)
+  - Tuple with optional type (`[number, string?]`)
+  - Tuple with spreads (`[...Array<number>, number, number]`)
+  - Tuple with optional type and spread (`[number?, ...Array<string>]`)
+  - Tuple with union values (`[number, string | undefined, boolean]`)
+  - Tuple with literal values (`[number, "cat" | "dog", true]`)
+- Records
+  - Record with string keys (`Record<string, number>`)
+  - Record with number keys (`Record<number, string>`)
+  - Record with union keys (`Record<string | number, unknown>`)
+  - Record with union values (`Record<string, string | number>`)
+  - Record with undefined values (`Record<string, string | undefined>`)
+  - Record with literal keys (`Record<"cat" | "dog", number>`)
+  - Record with literal values (`Record<string, 1 | 2>`)
+  - Record with template keys (``Record<`id_${number}`, string>``)
+  - Union of records (`Record<string, unknown> | Record<number, unknown>`)
+- Objects
+  - Object with named keys (`{ a: number }`)
+  - Object with union values (`{ a: string | number }`)
+  - Object with literal union values (`{ a: "cat" | 1 }`)
+  - Object with optional keys (`{ a?: number }`)
+  - Object with symbol keys (`{ [Symbol("a")]: number }`)
+  - Union of objects (`{ a: 1, b: 2 } | { b: 2, c: 3 }`)
+  - Nested objects (`{ a: { b: { c: 1 } } }`)
+- `readonly` versions of the above
+- `null` and `undefined`
+
+### Design philosophy
+
+**No type annotations.** Functions shouldn't require type annotations to have good types. This gives the best developer experience. The exception is `piped`.
+
+**Have good output types.** One reason Remeda provide "one-liner" functions is if we can have better output types than the defaults. While `data.filter(item => typeof item === 'number')` works, `filter(data, isNumber)` gives a better output type. This is why we have type guards, `hasAtLeast`, `prop`, etc.
```

Comment:
Everything in Remeda is "one-liner" so this sentence wasn't clear to me. We can mention that we provide alternative types for built-ins and trivial implementations for the sake of better typing, but even that doesn't cover the whole thing, we also provide implementations if we want to offer a lazy variant or sometimes just so that it could be used dataLast...

I think it's good to mention that output types are a major concern for the library, just think the phrasing and reasoning is different. We want to move errors from runtime to compile time, so better typing means more compile errors and less runtime errors.

Reply:
my idea behind this was that i wanted to write down our reasoning for why we have certain functions. i guess typing, data-last, and pipes are all reasons, which i might break into a different point (something around working well with the rest of remeda?)